compiler/src/iree/compiler/Codegen/Common/TransformExtensions/TransformMatchers.h

```diff
@@ -0,0 +1,459 @@
+           "null operation");
+    llvm::append_range(payloadOperations, operations);
+    payloadGroupLengths.push_back(payloadOperations.size() - originalLength);
+    return payloadGroupLengths.size() - 1;
+  }
+  void addPayloadGroup(ArrayRef<Operation *> operations) {
+    addPayloadGroup<ArrayRef<Operation *>>(operations);
+  }
+
+  /// Adds a new singleton list of payload operation to the list of lists if the
+  /// operation is non-null, adds an empty list otherwise. Useful for results of
+  /// optional matches.
+  void addPotentiallyEmptyPayloadGroup(Operation *op) {
+    if (!op)
+      addPayloadGroup(ArrayRef<Operation *>());
+    else
+      addPayloadGroup(ArrayRef<Operation *>(op));
+  }
+
+ private:
+  /// The flat list of all payload opreations. `payloadGroupLengths` can be used
+  /// to compute the sublist that corresponds to one nested list.
+  // TODO: if somebody implements such a flattened vector generically, use it.
+  SmallVector<Operation *> payloadOperations;
+  SmallVector<int64_t> payloadGroupLengths;
+};
+
+/// A transform state extension that maintains the mapping between callback
+/// names as strings usable in `match_callback` and their implementations.
+class MatchCallbacksRegistry : public transform::TransformState::Extension {
+ public:
+  using MatchCallbackFn = std::function<DiagnosedSilenceableFailure(
+      MatchCallbackResult &, Location, const transform::TransformState &,
+      ValueRange)>;
+
+  /// Constructs the extension.
+  MatchCallbacksRegistry(transform::TransformState &state)
+      : transform::TransformState::Extension(state) {}
+
+  /// Registers the given function as a callback with the given name. The name
+  /// must not be already present in the registry. The callback must be
+  /// convertible to MatchCallbackFn.
+  template <typename Fn>
+  void registerCallback(StringRef name, Fn &&fn) {
+    bool succeeded = callbacks.try_emplace(name, std::forward<Fn>(fn)).second;
+    (void)succeeded;
+    assert(succeeded && "adding a callback with a repeated name");
+  }
+
+  /// Returns a pointer to the implementation of the callback with the given
+  /// name, or null if it is not present in the registry.
+  const MatchCallbackFn *get(StringRef name) const {
+    auto iter = callbacks.find(name);
+    if (iter == callbacks.end()) return nullptr;
+    return &iter->getValue();
+  }
+
+ private:
+  llvm::StringMap<MatchCallbackFn> callbacks;
+};
+
+//===---------------------------------------------------------------------===//
+// Case-specific matcher builders.
+//===---------------------------------------------------------------------===//
+
+/// Creates a group of matchers for:
+///
+///     trailing(reduction(leading(), fill()))
+///
+/// where trailing and leading are elementwise operations whose presence is
+/// optional. Each matcher will capture the corresponding operation.
+void makeGPUReductionMatcher(StructuredOpMatcher &reduction,
```

Comment:
ditto, it looks like GPU specific method, so it should not live in Common/.

Reply:
Trivial followup move that will happen independently as discussed above.